### DISCUSSION TITLE
Adding custom tabs to Ilia (calculator)

### DISCUSSION THREAD

**Discussion by jnovek** [regolith-linux/regolith-desktop#675] [Q&A]
One nice thing about Rofi was that it was pluggable.

I had a back-burner project to write a simple Rofi calculator -- just a direct interface to `bc`.  Since I never got around to finishing that, I thought I might do the same thing with Ilia, but it looks like all the tabs (DialogPages?) are part of the monolithic application and written in Vala, which I've never used before.

It seems like right now I'd have to learn Vala (and a significant amount of GTK, it seems), fork Ilia, develop my math tab and hope that the maintainers consider a calculator a good addition to the tool when I send a PR.

I'm guessing y'all probably don't want to ship every nifty little tool that someone writes for ilia (I'm thinking: a color picker next) so what's the sustainable route look like?

Thanks!

Jason

**Comment by kgilmer** [comment]
Hi @jnovek , yes you're right in that `ilia` is a monolith.  However it was written with two extension mechanisms in mind:

1. Take as input a tree of menus with transitions and validations expressed in-line.  This is like an advanced form of `textlist` that allows for richer interaction.  It would not though be hostable via a shell script as is commonly done with `rofi`.
2. Plugins

So for option #2 there is already the basics for what a plugin API would need.  For example, there are a few interfaces that each page and the window use to communicate.  The missing piece is the plugin loading mechanism.  I did not want to add this feature until/if it was requested by someone.  Ideally the startup cost for such an addition in zero but usually there is some kind of cost, which is another reason it hasn't been added yet.  But to your other point, yes assuming there was a plugin architecture in place, you'd still need to learn some Vala, and also know some details about GTK / GDK.  Is that something you'd be interested in taking on in your quest for a pop-up calculator?

OTOH, Using `rofi` over `ilia` is supported in Regolith 2.  It's up to you to theme it as you like (they've removed support for Xresources recently) but you're still able to use Rofi to create your app.  You could even keep ilia for some things and rofi for others if you like.  The integration point is the i3 config file that specifies what app to launch when the given hotkey is pressed.
